cpp/include/cudf/dictionary/dictionary_column_view.hpp

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column.hpp>
+
+namespace cudf
+{
+/**
+ * @brief Given a column-view of DICTIONARY type, an instance of this class
+ * provides a wrapper on this compound column for dictionary operations.
+ *
+ * A dictionary column contains a dictionary and a set of indices.
+ * The dictionary is a sorted set of unique values for the column.
+ * The indices values are position indices into the dictionary.
+ */
+class dictionary_column_view : public column_view
+{
+public:
+    dictionary_column_view( column_view dictionary_column );
+    dictionary_column_view( dictionary_column_view&& dictionary_view ) = default;
+    dictionary_column_view( const dictionary_column_view& dictionary_view ) = default;
+    ~dictionary_column_view() = default;
+    dictionary_column_view& operator=(dictionary_column_view const&) = default;
+    dictionary_column_view& operator=(dictionary_column_view&&) = default;
+
+    /**
+     * @brief Returns the internal dictionary column.
+     */
+    column_view dictionary() const noexcept { return _dictionary; }
+
+    /**
+     * @brief Returns the column of indices
+     */
+    column_view indices() const;
+
+private:
+    column_view _dictionary;
+};
+
+namespace dictionary
+{
+
+/**
+ * @brief Create a new dictionary column by adding the new dictionary elements
+ * to the existing dictionary_column.
+ *
+ * The indices are updated if any of the new dictionary elements are sorted
+ * before any of the existing dictionary elements.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,1,0]}
+ * d2 = add_dictionary(d1,["b","c"])
+ * d2 is now {["a","b","c","d"],[3,0,2,0]}
+ * ```
+ *
+ * @throw cudf_logic_error if the dictionary type does not match the dictionary type in
+ * the dictionary_column.
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param dictionary New keys to incorporate into the dictionary_column
+ * @param mr Resource for allocating memory for the output.
+ * @return New dictionary column.
+ */
+std::unique_ptr<column> add_dictionary( dictionary_column_view const& dictionary_column,
```

Comment:
Would this name be clearer / reflect the behavior better?  Especially given the namespace, having dictionary in the name is not strictly necessary. (`cudf::dictionary::add_keys` vs. `cudf::dictionary::add_dictionary`).

```suggestion
std::unique_ptr<column> add_keys( dictionary_column_view const& dictionary_column,
```